front: merge ui and base prettier config and check ui formatting in ci

Synar opened the pull request:
ui formatting was not checked in the ci, as it was in the prettierignore of the base repo, and we didn't have a specific check for it. However, its config is pretty much identical to the base one, and having 2 prettierrc doesn't seem justified. 

So simply delete the prettierc of the ui folder, remove ui from the main prettierignore, and run prettier on everything at once.

emersion commented:
> it was in the prettierignore of the base repo

This was by design, to keep osrd-ui self-contained. Also we don't want to lint/touch any osrd-ui source files when building/linting/formatting osrd itself.

Math-R reviewed (approved):
lgtm

emersion reviewed (changes requested):
I don't think this is the right approach, sorry!

Synar commented:
> > it was in the prettierignore of the base repo
> 
> This was by design, to keep osrd-ui self-contained. 

Is there a single use case or practical benefit we currently gain from this, though? Is there ever a time where you make change to both the front and ui, and only want to format one of them?

There are clear practical disadvantages to the current approach though. First, if you want to format your work, you have to run 2 different commands, which is error prone. Second, the ci was clearly not formatting all files, and getting changes from running prettier on a repo formatted with prettier is unexpected for a dev.

> Also we don't want to lint/touch any osrd-ui source files when building/linting/formatting osrd itself.

This is not coherent with our linting configuration: 
```
 "lint": "npm run lint --workspaces --if-present && eslint --ext .ts,.tsx,.js,.jsx src scripts tests *.ts --max-warnings 0",
    "lint-fix": "npm run lint:fix --workspaces --if-present && eslint --ext .ts,.tsx,.js,.jsx src scripts tests *.ts --fix",
```

I don't see a reason why linting (which is pretty lengthy) should also lint workspaces, but formatting (which is much faster) should not.

Similarly build in osrd builds everything:
```"build": "npm run build --workspaces --if-present && vite build",```

emersion commented:
> Is there a single use case or practical benefit we currently gain from this, though? Is there ever a time where you make change to both the front and ui, and only want to format one of them?

Easier to figure out what's going on because files outside osrd-ui don't affect files inside, external contributors don't need to grok anything outside of `front/ui/`, better isolation between unrelated packages (and no cross-package dependencies).

> First, if you want to format your work, you have to run 2 different commands, which is error prone.

We can easily add a command to `package.json` which formats everything if need be, similar to the existing `lint` commands.

> the ci was clearly not formatting all files

That's a separate bug, which ought to be separately fixed.

> This is not coherent with our linting configuration

See above - this is the NPM script, but none of the tools we run in OSRD ever operate on both OSRD an osrd-ui (vite, eslint, prettier, tsc).

Synar commented:
Adding a npm format script (that also format ui by calling npm format in ui, like we do with lint) is fine with me, and solve pretty much the same issues as my current changes. So if you're fine with it, I'll sidestep the debate and do it that way ^^

Synar committed:
front, ui: update front and add ui prettier ignore

Signed-off-by: Alice Khoudli <alice.khoudli@polytechnique.org>

Synar committed:
ci: check ui formatting with prettier

This was in theory done through eslint for js/ts/tsx files,
but not for json or css files yet.

Signed-off-by: Alice Khoudli <alice.khoudli@polytechnique.org>

Synar committed:
front, ui: prettier check css and json in npm run lint

Signed-off-by: Alice Khoudli <alice.khoudli@polytechnique.org>

Synar commented:
Fully reworked the PR

emersion commented on front/ui/.prettierignore:
```diff
@@ -0,0 +1,2 @@
+**/*.md
+**/dist
```
Nit: the `**/` prefix is unnecessary here.

Synar replied:
Done ^^

emersion commented on .github/workflows/build.yml:
```diff
@@ -728,11 +728,11 @@ jobs:
           docker run --name=front-format --net=host \
             ${{ fromJSON(needs.build.outputs.stable_tags).front-tests }} \
-            npx prettier . --check
+            sh -c "npx prettier '**/*.{css,json}' --check && cd ui && npx prettier '**/*.{css,json}' --check"
```
I would prefer to just check all files instead of introducing yet another list of file extensions we need to keep in sync. (I don't feel strongly about this.)

flomonster replied:
I agree with @emersion.

Synar replied:
The point of adding file extensions is to remove this check entirely as soon as we support both these in eslint. It also speeds up the CI. I remember you told me you were opposed to checking the formatting of each file twice.

If we don't care about formatting twice, we could definitely remove these.

flomonster replied:
Let's add a comment to explain that it's temporary.

emersion replied:
@Synar Would you mind adding a TODO?

Synar replied:
Done ^^

emersion commented on front/package.json:
```diff
@@ -157,8 +157,8 @@
     "test": "npm run test --workspaces --if-present && vitest",
-    "lint": "npm run lint --workspaces --if-present && eslint --ext .ts,.tsx,.js,.jsx src scripts tests *.ts --max-warnings 0",
-    "lint-fix": "npm run lint:fix --workspaces --if-present && eslint --ext .ts,.tsx,.js,.jsx src scripts tests *.ts --fix",
+    "lint": "npm run lint --workspaces --if-present && eslint --ext .ts,.tsx,.js,.jsx src scripts tests *.ts --max-warnings 0 && prettier '**/*.{css,json}' --check && cd ui && prettier '**/*.{css,json}' --check",
```
Hm, putting the ui-related format checks here means that running `npm run lint` from inside an ui package will not run prettier. On the other hand, to fix it we'd need to add it to each and every `package.json` file… (We should be able to get rid of this with ESLint v9.)

Synar replied:
Indeed. But also the prettierrc is in the main ui folder and not each package, so it made sense to me to it this way. We could definitely add something like `npx prettier --config ../.prettierrc.json --ignorepath ../.prettierignore . --check` to each package.json, but it's a handful. Or duplicate each of the prettier config files inside the actual packages.

Since it's temporary, I went for the simplest option.

emersion replied:
Alright!

emersion reviewed (commented):
Thanks a lot! Here are a few comments, but overall this looks good!

Synar committed:
ui: fix prettier formatting issues

Signed-off-by: Alice Khoudli <alice.khoudli@polytechnique.org>

flomonster reviewed (approved):
LGTM

emersion commented on front/ui/storybook/package.json:
```diff
@@ -14,8 +14,8 @@
     "build": "storybook build",
-    "lint": "eslint stories .storybook --max-warnings 0",
-    "lint:fix": "eslint stories .storybook --fix"
+    "lint": "eslint stories .storybook --max-warnings 0 && prettier '**/*.{css,json}' --check --config ../.prettierrc.json --ignore-path ../.prettierignore",
```
Do we really need to manually specify `--config` and `--ignore-path`? AFAIK prettier should look into parent directories on its own when it doesn't find a config file in the current one.

Synar replied:
It does a poor job of it, unfortunately, which is why I went the explicit route. I'm not sure if it finds its config, but I'm almost certain it does not find its ignore file. I will double check again.

Synar replied:
Double checked, and it does find the config, but it indeed does not find the ignore file. This seems consistent with the doc.

emersion reviewed (approved):
Thanks!
